### DISCUSSION TITLE
Save image with background as png with write_to_buffer

### DISCUSSION THREAD

**Discussion by Delphilante** [libvips/libvips#4025] [Q&A]
When I use write_to_buffer as jpg, and set the background=255, the transparency disappears and gets the color as I want. However, the quality is awful, so I'd rather keep it as png. I can't figure out what kind of VipsArrayDouble I'm supposed to enter to remove the transparent background in png. I'm using Pyvips.

`base = base.write_to_buffer(f'{group}.jpg', background=55)` works
`base = base.write_to_buffer(f'{group}.png', background=55)` keeps the background transparent (same for `[255,255,255,255]` and anything else I've tried.

**Comment by jcupitt** [comment]
Helllo @Delphilante,

Use `flatten()` to remove alpha. You only need to pass the file extension to `write_to_buffer()` (it's used to select the saver).

```python
data = image.flatten().write_to_buffer('.png')
```

**Comment by Delphilante** [comment]
Thanks! One step closer to my goal ^^

Two questions:
- How am I supposed to color the flattened background now? It's now fixed to black.
- What is the point of having a parameter `background` if it doesn't do much for `pngsave`? I tried `vips pngsave` and `vips jpegsave`, both state they have the parameter `background`, yet the .png version somehow still won't accept my colors.

EDIT: Oh, figured it was inside `flatten`, I just forgot to state the argument (`flatten(background=50)`).
Second question remains, I wonder what the purpose is of having an argument `background` in `pngsave`.

**Reply by jcupitt** [reply]
The `background` param is defined on the `VipsForeignSave` base class, so it appears for all savers.

It's used in eg. `jpegsave` if you pass an image with an alpha, since JPEG (of course) does not support alpha channels. PNG supports both RGB and RGBA, so it doesn't make sense to use the background param.

**Reply by jcupitt** [reply]
Oh, you can do eg. `flatten(background=[100, 200, 0])`, you probably saw.
